core/src/main/scala/zio/config/ConfigDescriptorModule.scala

```diff
@@ -1206,6 +1207,177 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
             })
         )({ case (a, t) => a :: t }, l => l.headOption.map(h => (h, l.tail)))
 
+    /**
+     * enum allows user to up-cast all the subtypes to its super type defined by `D`.
+     * This is mainly useful in defining `coproducts` (`sealed trait`)
+     *
+     * Example:
+     * {{{
+     *   sealed trait D
+     *
+     *   case class A(a: String) extends D
+     *   case class B(b: Int) extends D
+     *   case class C(c: Double) extends D
+     *
+     *   val config: ConfigDescriptor[D] =
+     *     enum[D](
+     *       string("a")(A.Apply, A.unapply),
+     *       int("b")(B.apply, B.unapply),
+     *       double("c")(C.apply, C.unapply)
+     *     )
+     * }}}
+     *
+     * Currently enum supports to a maximum of 9 terms
+     *
+     * NOTE:
+     *
+     * Use zio-config-magnolia for better compile time safety when it comes to `sealed trait`,
+     * as it has strong compile time behaviour and makes sure all subtypes are being handled.
+     * On the other hand, `enum` doesn't complain at compile time if you forgot
+     * to pass the config descriptor of any of the subtype.
+     *
+     * Example:
+     *
+     * {{{
+     *   import zio.config.magnolia._
+     *
+     *   val config = descriptor[D]
+     * }}}
+     */
+    def enum[D] = new PartiallyAppliedEnum[D]
```

Comment:
We should call this `enumeration` rather than `enum`, because in Scala 3, `enum` is a keyword.